content/post/2024/02-sky130-kicad.md

```diff
@@ -0,0 +1,77 @@
+
+All joking aside, xschem is an incredibly powerful schematic capture system, but that power comes at a cost. xschem is fairly arcane to most, it has some graphical problems, and is just in general really hard to use if you're just starting out which makes rapid iteration difficult.
+
+Enter [KiCad], a robust and Open Source EDA software package, while it is mainly used for PCB layout, we can leverage the schematic capture and built-in SPICE simulation, right?
+
+## Introducing kicad-pdk-libs
+
+A little while ago I ran into [that exact problem](https://twitter.com/lethalbit/status/1498345181359321091), and so I set out and made [kicad-pdk-libs], a KiCad symbol library that adds schematic symbols for not only the basic building blocks of the sky130 PDK, but also SPICE library linkage.
+
+With this you can do things as simple as just stubbing out a SPICE simulation for some of the primitive gates, or go down to the transistor level and build gates from scratch and simulate their characteristics!
+
+![The KiCad "Choose Symbol" dialog showing a symbol called "sky130_nfet_01v8" being selected](/static-2024/sky130-kicad/kicad_nfet.png)
+
+As it stands, [kicad-pdk-libs] has almost 100% complete symbols for both sky130A and sky130B, for all the cell libraries in addition to the base `sky130_fd_pr` primitives library. This allows you to not just put transistor level schematics together, but much larger schematics using all of the pre-made gates and cells in the library, including some SRAM blocks.
+
+## Quick n' Dirty Inverter
+
+The basic example that almost everyone uses as their first CMOS device is the trusty inverter. As such we will quickly go over it here, if you wish for a more step-by-step guide with additional details, see the kicad-pdk-libs [intro] in the docs.
+
+All you need to do is to throw a `sky130_pfet_01v8` and a `sky130_nfet_01v8` on top of each other with their drains facing each other, hook up `VCC` and `GND` and then join the gates and drains, and just like that, you have an inverter using the primitive sky130 FET models.
+
+![The CMOS inverted as described above](/static-2024/sky130-kicad/inverter_sch.png)
+
+Next setup some way to drive the sim, in my case I did a VDC source and then a VPULSE to drive the inverter, these are built-in to the standard KiCad SPICE library, it should look something like this when done:
+
+![Full view of the KiCad schematic showing the VDC and VPULSE elements along with the CMOS inverter previously shown](/static-2024/sky130-kicad/schema.png)
+
+After that, you can add the following SPICE directives as a text element on the schematic, and make sure to replace `${PDK_ROOT}` with the path to your local PDK root:
+
+```spice
+.tran 100f 100n
+.lib ${PDK_ROOT}/sky130A/libs.tech/ngspice/sky130.lib.spice tt
```

Comment:
wow! I learnt something super useful today! thanks! .. Oh dang, this is only for Kicad? I was hoping it would work for ngspice in the shell as well! I always have to hard code the pdk path

Reply:
Yeah sadly this is only for KiCad, I was looking for a way to do variable expansion in pure ngspice but i've not found a way sadly.